clab/clab.go

```diff
@@ -527,22 +542,109 @@ func (c *CLab) scheduleNodes(ctx context.Context, maxWorkers int,
 					continue
 				}
 
+				// we meed to populate e.g. the mgmt net ip addresses before continuing with the
+				// post-deploy phase (for e.g. certificate creation)
+				err = node.UpdateConfigWithRuntimeInfo(ctx)
+				if err != nil {
+					log.Errorf("failed to update node runtime information for node %s: %v", node.Config().ShortName, err)
+				}
+
+				dm.SignalDone(node.GetShortName(), types.WaitForCreate)
+
+				err = dm.Enter(node.GetShortName(), types.WaitForCreateLinks)
+				if err != nil {
+					log.Error(err)
+				}
+
 				err = node.DeployLinks(ctx)
 				if err != nil {
 					log.Errorf("failed deploy links for node %q: %v", node.Config().ShortName, err)
 					continue
 				}
 
-				// signal to dependency manager that this node is done with creation
-				dm.SignalDone(node.Config().ShortName, dependency_manager.NodeStateCreated)
+				// DeployLinks does not necessarily create all the links already.
+				// veth links might be created by the peer node, if it was not already
+				// deployed properly. Hence we need to wait for all the links to be created.
+				// we just wait if there is actually a dependency on this state, otherwise
+				// we head on.
```

Comment:
everything below this block makes the function very long
let's try to carve it to decompose it

Reply:
resonable, yes

Reply:
postponing for a later enhancements

Reply:
Agreed to be tackled in an upcoming PR.